chore: add lint-pr action

ryanchristo opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

This pull request adds the configuration for a "Lint PR" action (see [action-semantic-pull-request](https://github.com/amannn/action-semantic-pull-request)) that ensures each pull request title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification.

**This update will require future pull request titles to be prefixed with one of the [default types](https://github.com/commitizen/conventional-commit-types/blob/master/index.json) before the pull request can be merged. There is no requirement for scope at this time.**

ref #9268

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md). - n/a
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing) - n/a
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`) - n/a
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code). - n/a
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md` - n/a
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

ryanchrypto commented on .github/workflows/lint-pr.yml:
```diff
@@ -0,0 +1,52 @@
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          # Configure which types are allowed.
+          # Default: https://github.com/commitizen/conventional-commit-types
+          # types: |
+          #   fix
+          #   feat
```
Should we use the [default commit types](https://github.com/commitizen/conventional-commit-types/blob/master/index.json) or should we limit the commit types to the same types that we end up choosing for pull request templates (see #9341)?

marbar3778 replied:
It would be nice to prefix with what the change touches. If the pr is working on staking then a prefix of `staking: <subject>` is my preferred approach. I also dont follow the need for so many templates that all say the same thing?

ryanchristo replied:
> It would be nice to prefix with what the change touches.

This is the purpose of the scope. The type is what type of change is being made (i.e. feat, fix, refactor, chore, etc.). Here is an example from [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/):

```
feat(parser): add ability to parse arrays
```

> I also dont follow the need for so many templates that all say the same thing?

This could still use some fine-tuning and maybe consolidating a couple types into the same template (e.g. all the types that do not change production code). The idea with separate pull request templates is to make the checklist more relevant to each pull request (e.g. we only need a check for following the building modules documentation for feature pull requests).

aaronc replied:
I think the default is a good starting point

ryanchrypto commented on .github/workflows/lint-pr.yml:
```diff
@@ -0,0 +1,52 @@
+          # types: |
+          #   fix
+          #   feat
+          # Configure which scopes are allowed.
+          # scopes: |
+          #   core
+          #   ui
+          # Configure that a scope must always be provided.
+          # requireScope: true
```
I will remove these comments unless we think limiting the scopes is necessary.

ryanchristo replied:
The scope will not be required. Using default configuration for now.

ryanchrypto commented:
cc @aaronc @alexanderbez @marbar3778

Any initial thoughts? Is this something we should discuss on one of our community calls first? Also, I can remove the commented out configuration options but I left them here initially for visibility.

marbar3778 commented:
I'm a bit lost on what types and scope actually mean here? Is there an example of how the PR would look with this?

ryanchristo commented:
> I'm a bit lost on what types and scope actually mean here? Is there an example of how the PR would look with this?

The type is what type of change is being made (i.e. feat, fix, refactor, chore, etc.). The scope is to further specify the topic of the change (e.g. what package the changes touch). Here is an example from [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/):

```
feat(parser): add ability to parse arrays
```

The general idea is to improve how we track changes and to further encourage atomic pull requests. Here is a snippet from the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification:

> It provides an easy set of rules for creating an explicit commit history; which makes it easier to write automated tools on top of. This convention dovetails with SemVer, by describing the features, fixes, and breaking changes made in commit messages.

marbar3778 commented:
oh okay, Thanks for the explanation. I think it makes sense to be this strict.

alexanderbez commented:
I'm certainly in favor of something like this, but two initial questions:

1. What does it matter since all commits are squashed anyway and only the final commit matters?
2. How will this play with [Go's convention](https://golang.org/doc/contribute#commit_messages) re commit messages that specifics the prefix should be the package and not a type?

ryanchristo commented:
> What does it matter since all commits are squashed anyway and only the final commit matters?

From the [action-semantic-pull-request](https://github.com/amannn/action-semantic-pull-request) description, "This is a Github Action that ensures that your PR title matches the Conventional Commits spec." This action is not concerned with individual commits, it only checks the PR title, which will become the commit when all the commits have been squashed and merged.

> How will this play with Go's convention re commit messages that specifics the prefix should be the package and not a type?

The Go convention proposes to prefix the first line of the commit with the package name. Conventional Commits proposes to use the type of change and a scope (which could be the package name). Conventional Commits is more specific and extends beyond the Go community, but you are right, these two conventions are not aligned and we should decide on one or the other.

One of the advantages of using [action-semantic-pull-request](https://github.com/amannn/action-semantic-pull-request) is that it works with tools like [semantic-release](https://github.com/semantic-release/semantic-release). At Regen Network, we have been discussing ways to improve upon the release process and this convention for commits was one suggestion. Also, another example of this in use is the [dependapot pull requests](https://github.com/cosmos/cosmos-sdk/pulls?q=is%3Apr+author%3Aapp%2Fdependabot), so in some ways we are already following this convention.

robert-zaremba commented:
The idea is good.

I have one concern: Will this work with our auto-squashing / automerge functionality?

aaronc commented:
> The idea is good.
> 
> I have one concern: Will this work with our auto-squashing / automerge functionality?

Yes as long as the check passes

aaronc commented:
I just want to clarify one goal I have related to this is the idea that we have QA checklists for PR reviewers that are related to the type of PR. So a docs PR would have a different QA checklist than a feature PR.

Now, conventional commits is not the only way to achieve this. But, it is a somewhat widely used standard and focuses the PR on one and only one type of change (docs, feature, fix, etc.) which is generally good. We could alternatively use `Type: ` labels but it's harder to enforce one label per PR.

Regarding PR scope, which could be indicated with the golang convention (ex. `math: ...`) or conventional commits (ex. `feat(parser): ...`), I think it's harder to limit PR's to a single package/module than a single _type_ of change. In particular, "refactor" PRs generally span many if not all modules. Also, we have `C:` (for component) labels to indicate PR scope pretty well and we have the automatic labeler action to tag PRs. To generally be able to limit PRs to a single package, we would probably need to split the SDK into many separate go modules which is not a bad idea IMHO but beyond the scope of what we're trying to do here.

So, basically I support these conventional commit prefix tags (`feat:`, `fix:`, etc.) as a good way to focus PRs on one type of change and have QA checklists accordingly. I prefer not including scope (ex. `feat(parser):`) for now and just using labels for that, but I'd be open to it later if we refactor the SDK enough to make it easy to limit PRs to one type of change + one component.

aaronc commented on .github/workflows/lint-pr.yml:
```diff
@@ -0,0 +1,52 @@
+          # will suggest using that commit message instead of the PR title for the
+          # merge commit, and it's easy to commit this by mistake. Enable this option
+          # to also validate the commit message for one commit PRs.
+          validateSingleCommit: true
```
Do we want/need this?

ryanchristo replied:
It might not be necessary. We should probably just start with the default configuration. I will update.

ryanchristo commented:
I've update the configuration to use the default configuration. See [configuration](https://github.com/amannn/action-semantic-pull-request#configuration) for additional configuration options.

ryanchristo commented:
Also, updated pull request summary and added link to [default types](https://github.com/commitizen/conventional-commit-types/blob/master/index.json) for reference.

blushi reviewed (approved):
LGTM
Just an internal note: it would be good to set this up too on regen-ledger

Jsn2win reviewed (commented):
pr lint
